cpp/src/barretenberg/srs/io.hpp

```diff
@@ -17,28 +18,381 @@ struct Manifest {
     uint32_t start_from;
 };
 
-void read_transcript_g1(g1::affine_element* monomials, size_t degree, std::string const& dir);
+// Detect whether a curve has a G2AffineElement defined
+template <typename Curve> concept HasG2 = requires
+{
+    typename Curve::G2AffineElement;
+};
+
+// If Curve has a G2AffineElement type, check whether T is this type.
+template <typename Curve, typename T>
+concept GivingG2AffineElementType = std::same_as<T, typename Curve::G2AffineElement>;
+
+// If Curve has a G2AffineElement type, check whether T is this type.
+template <typename Curve, typename T>
+concept GivingG1AffineElementType = std::same_as<T, typename Curve::AffineElement>;
+
+template <typename Curve> class IO {
+    using Fq = typename Curve::BaseField;
+    using Fr = typename Curve::ScalarField;
+    using Element = typename Curve::Element;
+    using AffineElement = typename Curve::AffineElement;
+
+    static constexpr size_t BLAKE2B_CHECKSUM_LENGTH = 64;
+
+    static size_t get_transcript_size(const Manifest& manifest)
+    {
+        const size_t manifest_size = sizeof(Manifest);
+        const size_t g1_buffer_size = sizeof(Fq) * 2 * manifest.num_g1_points;
+        size_t result = manifest_size + g1_buffer_size + BLAKE2B_CHECKSUM_LENGTH;
+
+        if constexpr (HasG2<Curve>) {
+            const size_t g2_buffer_size = 2 * sizeof(Fq) * 2 * manifest.num_g2_points;
+            info(g2_buffer_size);
+            result += g2_buffer_size;
+        }
+
+        return result;
+    }
+
+    static void read_manifest(std::string const& filename, Manifest& manifest)
+    {
+        std::ifstream file;
+        file.open(filename, std::ifstream::binary);
+        file.read((char*)&manifest, sizeof(Manifest));
```

Comment:
You are not checking that enough data has been read here. If the file is small you might start accessing uninitialized memory